apache/metron
# METRON-1642: KafkaWriter should be able choose the topic from a field in addition to topology construction time

## Contributor Comments
Currently, we choose the kafka topic via the kafka.topic field.  It would be useful to allow people to specify the topic via a field.  This would enable multi-stage (or chain) parsing, among other use-cases.

## Manual Test
Ensure that `ZOOKEEPER`, `BROKERLIST` and `METRON_HOME` are set
* Create a topics for 2 sensors:
```
/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper $ZOOKEEPER --create --topic 1642_pre --partitions 1 --replication-factor 1
/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper $ZOOKEEPER --create --topic 1642_second --partitions 1 --replication-factor 1
```
* Open `$METRON_HOME/config/zookeeper/parsers/1642_pre.json` and save
```
{
  "parserClassName" : "org.apache.metron.parsers.csv.CSVParser"
  ,"sensorTopic" : "1642_pre"
  , "parserConfig": {
     "columns" : { "name": 0
                 , "first_field": 1
                 , "timestamp" : 2
                 },
     "kafka.topicField" : "route_field"
   }
  ,"fieldTransformations" : [
    {
     "transformation" : "STELLAR"
    ,"input" :  ["name"]
    ,"output" :  ["route_field"]
    ,"config" : {
      "route_field" : "match{ name == 'metron' => '1642_second', default => 'enrichments'}"
                }
    }
                           ]

}
```
* Open `$METRON_HOME/config/zookeeper/parsers/1642_second.json` and save
```
{
  "parserClassName" : "org.apache.metron.parsers.json.JSONMapParser"
  ,"sensorTopic" : "1642_second"
  , "parserConfig": {
   }
  ,"fieldTransformations" : [
    {
     "transformation" : "STELLAR"
    ,"input" :  ["first_field"]
    ,"output" :  ["new_field"]
    ,"config" : {
      "new_field" : "TO_UPPER(first_field)"
                }
    }
                           ]
}
```
* Push configs via `$METRON_HOME/bin/zk_load_configs.sh --mode PUSH -z $ZOOKEEPER -i $METRON_HOME/config/zookeeper/`
* Stop the parsers in ambari
* Start the `1642_pre` and `1642_second` parsers
```
$METRON_HOME/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s 1642_pre
$METRON_HOME/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s 1642_second
```
* Open `~/data.json` and save:
```
notmetron,f1,1
metron,f2,2
metron,f3,3
metron,f4,4
metron,f5,5
metron,f6,6
metron,f7,7
notmetroneither,f8,8
```
* Send data into the `1642_pre` topic via
```
cat data.json | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic 1642_pre
```

From here you should see data flowing into the `1642_pre` and `1642_second` indices depending on the first field.  When the field is `metron`, then it should go into the `1642_second` index and a new field `new_field` should be created.  Otherwise, it will go in as-is in the `1642_pre` index.
## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
